protected/humhub/modules/user/models/User.php

```diff
@@ -935,4 +940,28 @@ public function setMustChangePassword($state = true)
         }
     }
 
+    public static function getStatusOptions(bool $withDeleted = true): array
+    {
+        $options = [
+            self::STATUS_ENABLED => Yii::t('AdminModule.user', 'Enabled'),
+            self::STATUS_DISABLED => Yii::t('AdminModule.user', 'Disabled'),
+            self::STATUS_NEED_APPROVAL => Yii::t('AdminModule.user', 'Unapproved'),
+        ];
+
+        if ($withDeleted) {
+            $options[self::STATUS_SOFT_DELETED] = Yii::t('AdminModule.user', 'Deleted');
+        }
+
+        return $options;
+    }
+
+    public static function getVisibilityOptions(): array
+    {
+        return [
+            self::VISIBILITY_REGISTERED_ONLY => Yii::t('AdminModule.user', 'Visible for members only'),
+            self::VISIBILITY_ALL => Yii::t('AdminModule.user', 'Visible for members+guests'),
```

Comment:
In previous versions, we did not show the "Guest" option if guest access was not enabled. Otherwise, there could be some confusion...

Reply:
Not sure the default profile visibility "hidden" should be possible.

Reply:
> In previous versions, we did not show the "Guest" option if guest access was not enabled. Otherwise, there could be some confusion...

Sorry, I don't find where in previous version we hide the option "Visible for members+guests", or what option you mean "Guest" and where I can find it?

> Not sure the default profile visibility "hidden" should be possible.

Do you mean to hide the option "Invisible" from here?

![hidden](https://user-images.githubusercontent.com/6995524/160649071-17744418-9fb0-4869-99bb-a266892a7372.png)

Reply:
@luke- I have checked the cases and done changes in commit https://github.com/humhub/humhub/pull/5607/commits/b10b9d7952d9f7a3faf2e06ae5a8e80977e9a0f9:

> Hidden users are not be visible for regular users. e.g. at

- People Directory +
- Space Memberslist / Sidebar Snippet +
- Activities sidebar - *Done in the commit*
- Space About Site +
- As Followers/Friends +
- Search - *unfortunately I could not find a solution yet, if you know please tell me, or I will do this later*
- Mentioning - *Done in the commit*

> They will still be displayed in the following places:

- User Administration +
- In the group administration (members) +
- In the Space Members administration + Owner Change Form +
- REST API +

Reply:
> > In previous versions, we did not show the "Guest" option if guest access was not enabled. Otherwise, there could be some confusion...
> 
> Sorry, I don't find where in previous version we hide the option "Visible for members+guests", or what option you mean "Guest" and where I can find it?

You need to enable Guest Access, then the Dropdown appear.

> 
> > Not sure the default profile visibility "hidden" should be possible.
> 
> Do you mean to hide the option "Invisible" from here?
> 
> ![hidden](https://user-images.githubusercontent.com/6995524/160649071-17744418-9fb0-4869-99bb-a266892a7372.png)

Yes

Reply:
> @luke- I have checked the cases and done changes in commit [b10b9d7](https://github.com/humhub/humhub/commit/b10b9d7952d9f7a3faf2e06ae5a8e80977e9a0f9):
> 
> > Hidden users are not be visible for regular users. e.g. at
> 
> * People Directory +
> * Space Memberslist / Sidebar Snippet +
> * Activities sidebar - _Done in the commit_
> * Space About Site +
> * As Followers/Friends +
> * Search - _unfortunately I could not find a solution yet, if you know please tell me, or I will do this later_
> * Mentioning - _Done in the commit_
> 
> > They will still be displayed in the following places:
> 
> * User Administration +
> * In the group administration (members) +
> * In the Space Members administration + Owner Change Form +
> * REST API +

Can you maybe add some acceptance tests to that?

Reply:
@luke- 

> > > Not sure the default profile visibility "hidden" should be possible.

> > Do you mean to hide the option "Invisible" from here?

> Yes

Commit https://github.com/humhub/humhub/pull/5607/commits/b32864811b815c140ab6d4ce3f045db0dde684a5

Reply:
@luke- 

> Can you maybe add some acceptance tests to that?

Commit https://github.com/humhub/humhub/pull/5607/commits/175ffbc5c18cda470308e3576cf9a085e66dc637

Reply:
@luke- 

>> Hidden users are not be visible for regular users. e.g. at
>
> - Search - **unfortunately I could not find a solution yet, if you know please tell me, or I will do this later**

Sorry, but I cannot find a solution to implement this, maybe you know how the `ZendLuceneSearch` works.
What I tried to do:

- in the `humhub\modules\search\engine\Search::getMetaInfoArray(Searchable $obj)` I added:

```php
} elseif ($meta['type'] == self::DOCUMENT_TYPE_USER && $obj->visibility != User::VISIBILITY_HIDDEN) {
    $meta['visibility'] = self::DOCUMENT_VISIBILITY_VISIBLE;
}
```

`const DOCUMENT_VISIBILITY_VISIBLE = 'visible';`

- then in the `humhub\modules\search\engine\ZendLuceneSearch::buildQuery()` added:

```php
if (isset($options['model']) && $options['model'] === User::class) {
    $userQuery = new Boolean();
    $userQuery->addSubquery(new QueryTerm(new Term(self::DOCUMENT_VISIBILITY_VISIBLE, 'visibility')), true);
    $userQuery->addSubquery(new QueryTerm(new Term(User::class, 'containerModel')), true);
    $query->addSubquery($userQuery, true);
}
```

I tried many variants of the `addSubquery()`, I played with the second param between `true`, `false`, `null`, also I tried with `new MultiTerm()` and removed the condition `if (isset($options['model']) && $options['model'] === User::class) {` and many others, but still cannot find a solution to exclude only hidden users.
It would be good if you know how the queries work.
Thanks.

Reply:
@yurabakhtin Can't we just not include Invisible users in the search index? Apparently there is already something similar here: 

https://github.com/humhub/humhub/blob/master/protected/humhub/modules/user/models/User.php#L567

Reply:
@luke- Thank you for the advice, but it didn't solve the issue, please see my change https://github.com/humhub/humhub/pull/5607/commits/b9d4b8de723451f63abbfcbe878306f31d189a85. I decided to do the change right inside the method `User->isVisible()` because it is used only in the method `User->updateSearch()` and the method's comment is `Specifies whether the user should appear in user lists or in the search.`.

I tested this change with running `php yii queue/listen`, I also run `php yii cron/run`(log was `Optimizing search index...` there), I updated a tested user many times, but all this didn't help to exclude a hidden user from search results.
Could you test this or do you have another idea? Thanks.